component/newstorage/batchedstore/batchedstore.go

```diff
@@ -0,0 +1,361 @@
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	value, err := s.underlyingStore.Get(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get value from underlying store: %w", err)
+	}
+
+	return value, nil
+}
+
+func (s *store) GetTags(key string) ([]newstorage.Tag, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	tags, err := s.underlyingStore.GetTags(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tags from underlying store: %w", err)
+	}
+
+	return tags, nil
+}
+
+func (s *store) GetBulk(keys ...string) ([][]byte, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	values, err := s.underlyingStore.GetBulk(keys...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get values from underlying store: %w", err)
+	}
+
+	return values, nil
+}
+
+func (s *store) Query(expression string, options ...newstorage.QueryOption) (newstorage.Iterator, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	iterator, err := s.underlyingStore.Query(expression, options...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get values from underlying store: %w", err)
+	}
+
+	return iterator, nil
+}
+
+func (s *store) Delete(key string) error {
+	if key == "" {
+		return errEmptyKey
+	}
+
+	s.Lock()
+
+	s.currentBatch = append(s.currentBatch, newstorage.Operation{
+		Key:   key,
+		Value: nil,
+	})
+
+	s.Unlock()
+
+	s.RLock()
+	currentBatchLength := len(s.currentBatch)
+	s.RUnlock()
+
+	if currentBatchLength >= s.batchSizeLimit {
+		err := s.Flush()
+		if err != nil {
+			return fmt.Errorf(failFlush, err)
+		}
+	}
+
+	return nil
+}
+
+func (s *store) Batch(operations []newstorage.Operation) error {
+	for _, operation := range operations {
+		if operation.Key == "" {
+			return errors.New("an operation's key was empty")
+		}
+	}
+
+	for _, operation := range operations {
+		s.Lock()
+
+		s.currentBatch = append(s.currentBatch, newstorage.Operation{
+			Key:   operation.Key,
+			Value: operation.Value,
+			Tags:  operation.Tags,
+		})
+
+		s.Unlock()
+
+		s.RLock()
+		currentBatchLength := len(s.currentBatch)
+		s.RUnlock()
+
+		if currentBatchLength >= s.batchSizeLimit {
+			err := s.Flush()
+			if err != nil {
+				return fmt.Errorf(failFlush, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *store) Close() error {
+	s.Lock()
+
+	s.close(s.name)
+
+	s.Unlock()
+
+	err := s.Flush()
+	if err != nil {
+		return fmt.Errorf(failFlush, err)
+	}
+
+	err = s.underlyingStore.Close()
+	if err != nil {
+		return fmt.Errorf("failed to close underlying store: %w", err)
+	}
+
+	return nil
+}
+
+func (s *store) Flush() error {
+	s.Lock()
+
+	var timeToFlush time.Duration
+
+	if len(s.currentBatch) > 0 {
+		startFlushTime := time.Now()
+
+		err := s.underlyingStore.Batch(s.currentBatch)
+		if err != nil {
+			return fmt.Errorf("failure while executing batched operations: %w", err)
```

Comment:
Need to unlock - or use defer Unlock() above.

Reply:
Oops. Done. I decided to use an unlock instead of defer so that the logging at the bottom of the function won't unnecessarily delay the release of the lock.